File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.cc
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <queue>
#include <algorithm>

using namespace std;
int N, M;
string arr[60];
bool visited[1 << 7][60][60] = { false, };
int start_y, start_x;
int moveY[4] = { -1,0,1,0 };
int moveX[4] = { 0,-1,0,1 };

int bfs(int y, int x) {
	queue<pair<pair<int, int>, pair<int, int> > > q;
	q.push(make_pair(make_pair(y, x), make_pair(0, 0)));
	visited[0][y][x] = true;

	while (!q.empty()) {
		int now_y = q.front().first.first;
		int now_x = q.front().first.second;
		int now_moving = q.front().second.first;
		int now_key = q.front().second.second;
		q.pop();

		if (arr[now_y][now_x] == '1') {
			return now_moving;
		}

		for (int i = 0; i < 4; i++) {
			int next_y = now_y + moveY[i];
			int next_x = now_x + moveX[i];
			if ((next_y >= 0) && (next_y < N) && (next_x >= 0) && (next_x < M) && (arr[next_y][next_x] != '#') && (!visited[now_key][next_y][next_x])) {
				if ((arr[next_y][next_x] == '.') || (arr[next_y][next_x] == '0') || (arr[next_y][next_x] == '1')) {
					visited[now_key][next_y][next_x] = true;
					q.push(make_pair(make_pair(next_y, next_x), make_pair(now_moving + 1, now_key)));
				}
				else if ((arr[next_y][next_x] >= 'a') && (arr[next_y][next_x] <= 'f')) {
					int tmp = now_key | (1 << (int(arr[next_y][next_x]) - 97));
					if (!visited[tmp][next_y][next_x]) {
						visited[now_key][next_y][next_x] = true;
						visited[tmp][next_y][next_x] = true;
						q.push(make_pair(make_pair(next_y, next_x), make_pair(now_moving + 1, tmp)));
					}
				}
				else if ((arr[next_y][next_x] >= 'A') && (arr[next_y][next_x] <= 'F')) {
					int tmp = int(arr[next_y][next_x]) - 65;
					if ((now_key & (1 << tmp))) {
						visited[now_key][next_y][next_x] = true;
						q.push(make_pair(make_pair(next_y, next_x), make_pair(now_moving + 1, now_key)));
					}
				}
			}

		}
	};

	return -1;
}

int main() {
	cin.tie(NULL);
	cout.tie(NULL);
	ios::sync_with_stdio(false);

	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		cin >> arr[i];
		for (int j = 0; j < M; j++) {
			if (arr[i][j] == '0') {
				start_y = i;
				start_x = j;
			}
		}
	}
	cout << bfs(start_y, start_x) << "\n";

	return 0;
}

File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.java
import java.io.*;
import java.util.*;

public class Main {
	
	private static class Position {
		
		int Y, X, K, C;

		public Position(int y, int x, int k, int c) {
			super();
			Y = y;
			X = x;
			K = k;
			C = c;
		}

		@Override
		public String toString() {
			return "Position [Y=" + Y + ", X=" + X + ", K=" + K + ", C=" + C + "]";
		}
		
	}
	
	private static int N, M;
	private static Position start;
	private static char[][] map = new char[51][51];
	private static boolean[][][] visited = new boolean[51][51][1 << 6];
	private static Queue<Position> queue = new LinkedList<>();
	private static int[] moveY = { -1,0,1,0 };
	private static int[] moveX = { 0,-1,0,1 };
	private static int Answer = 0;
    
    private static int bfs() {
    	queue.add(new Position(start.Y, start.X, 0, 0));
    	visited[start.Y][start.X][0] = true;
    	
    	while (!queue.isEmpty()) {
    		Position now = queue.poll();
    		int nowY = now.Y;
    		int nowX = now.X;
    		int nowKey = now.K;
    		int nowCount = now.C;
    		
    		if (map[nowY][nowX] == '1') {
    			return nowCount;
    		}
    		
    		for (int i = 0; i < 4; i++) {
    			int nextY = nowY + moveY[i];
    			int nextX = nowX + moveX[i];
    			if ((nextY >= 0) && (nextY < N) && (nextX >= 0) && (nextX < M) && !visited[nextY][nextX][nowKey] && (map[nextY][nextX] != '#')) {
        			char next = map[nextY][nextX];
    				if ((next >= 'a') && (next <= 'f')) {
    					int nextKey = nowKey | (1 << (next - 'a'));
    					visited[nextY][nextX][nextKey] = true;
    					queue.add(new Position(nextY, nextX, nextKey, nowCount + 1));
    				} else if ((next >= 'A') && (next <= 'F')) {
    					int isGetKey = nowKey & (1 << (next - 'A'));
    					if (isGetKey > 0) {
    						visited[nextY][nextX][nowKey] = true;
    						queue.add(new Position(nextY, nextX, nowKey, nowCount + 1));
    					}
    				} else if ((next == '.') || (next == '0') || (next == '1')) {
						visited[nextY][nextX][nowKey] = true;
						queue.add(new Position(nextY, nextX, nowKey, nowCount + 1));
    				}
    			}
    		}
    	}
    	
    	return -1;
    }
	
	private static void settings() {
		Answer = bfs();
	}
	
	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] inputs = br.readLine().split(" ");
        N = Integer.parseInt(inputs[0]);
        M = Integer.parseInt(inputs[1]);

        for (int i = 0; i < N; i++) {
        	String input = br.readLine();
        	for (int j = 0; j < M; j++) {
        		map[i][j] = input.charAt(j);
        		if (map[i][j] == '0') {
        			start = new Position(i, j, 0, 0);
        		}
        	}
        }
        settings();
        
        System.out.println(Answer);
        br.close();
    }

}